chia/util/action_scope.py

```diff
@@ -0,0 +1,174 @@
+class SideEffects(Protocol):
+    def __bytes__(self) -> bytes: ...
+
+    @classmethod
+    def from_bytes(cls: Type[_T_SideEffects], blob: bytes) -> _T_SideEffects: ...
+
+
+_T_SideEffects = TypeVar("_T_SideEffects", bound=SideEffects)
+
+
+@dataclass
+class ActionScope(Generic[_T_SideEffects]):
+    """
+    The idea of a wallet action is to map a single user input to many potentially distributed wallet functions and side
+    effects. The eventual goal is to have this be the only connection between a wallet type and the WSM.
+
+    Utilizes a "resource manager" to hold the state in order to take advantage of rollbacks and prevent concurrent tasks
+    from interferring with each other.
+    """
+
+    _resource_manager: ResourceManager
+    _side_effects_format: Type[_T_SideEffects]
+    _callbacks: List[Callable[[StateInterface[_T_SideEffects]], Awaitable[None]]] = field(default_factory=list)
+    _final_side_effects: Optional[_T_SideEffects] = field(init=False, default=None)
+
+    @property
+    def side_effects(self) -> _T_SideEffects:
+        if self._final_side_effects is None:
+            raise RuntimeError(
+                "Can only request ActionScope.side_effects after exiting context manager. "
+                "While in context manager, use ActionScope.use()."
+            )
+
+        return self._final_side_effects
+
+    @classmethod
+    @contextlib.asynccontextmanager
+    async def new_scope(
+        cls,
+        side_effects_format: Type[_T_SideEffects],
+        resource_manager_backend: Type[ResourceManager] = SQLiteResourceManager,
+    ) -> AsyncIterator[ActionScope[_T_SideEffects]]:
+        async with resource_manager_backend.managed(side_effects_format()) as resource_manager:
+            self = cls(_resource_manager=resource_manager, _side_effects_format=side_effects_format)
+            try:
+                yield self
+            except Exception:
+                raise
+            else:
+                async with self.use(_callbacks_allowed=False) as interface:
+                    for callback in self._callbacks:
+                        await callback(interface)
+                    self._final_side_effects = interface.side_effects
+
+    @contextlib.asynccontextmanager
+    async def use(self, _callbacks_allowed: bool = True) -> AsyncIterator[StateInterface[_T_SideEffects]]:
+        async with self._resource_manager.use():
+            memos = await self._resource_manager.get_memos()
+            side_effects = await self._resource_manager.get_resource(self._side_effects_format)
+            interface = StateInterface(memos, side_effects, _callbacks_allowed)
+            try:
+                yield interface
+            except Exception:
+                raise
+            else:
+                await self._resource_manager.save_memos(interface.memos)
+                await self._resource_manager.save_resource(interface.side_effects)
+                self._callbacks.extend(interface._new_callbacks)
```

Comment:
```suggestion
            yield interface

            await self._resource_manager.save_memos(interface.memos)
            await self._resource_manager.save_resource(interface.side_effects)
            self._callbacks.extend(interface._new_callbacks)
```

Reply:
Interestingly pylint complains about this one.  I don't fully understand it tbh: `chia\util\action_scope.py:142:12: W0135: The context used in function 'new_scope' will not be exited. (contextmanager-generator-missing-cleanup)`

Reply:
Changed back pending pylint fix in main: 94b2d65f53